rust-src/aggregate_sig/src/ffi.rs

```diff
@@ -137,6 +152,74 @@ pub extern "C" fn bls_verify_aggregate(
     u8::from(verify_aggregate_sig_trusted_keys(m_bytes, &pks, *sig))
 }
 
+/// Verify a an aggregate signature by verifying the groupings which consists of
+/// a message and public keys.
+#[no_mangle]
+#[allow(clippy::not_unsafe_ptr_arg_deref)]
+pub extern "C" fn bls_verify_aggregate_hybrid(
+    m_ptr: *const *const u8,        // array of pointers to messages.
+    message_lengths: *const size_t, // array of the lengths of the messages.
+    pks_ptr: *const *const *mut PublicKey<Bls12>, /* array of pointers to arrays of pointers
+                                     * to public keys */
+    pks_len: *const size_t, // number of public keys.
+    len: size_t,            /* the number of sets of public keys. Note. this is equal to the
+                             * number of messages. */
+    sig_ptr: *const Signature<Bls12>, // pointer to the signature.
+) -> u8 {
+    let m_lens_: &[size_t] = if len == 0 {
+        &[]
+    } else {
+        unsafe { slice::from_raw_parts(message_lengths, len) }
+    };
+
+    let ms_: &[*const u8] = if len == 0 {
+        &[]
+    } else {
+        unsafe { slice::from_raw_parts(m_ptr, len) }
+    };
+
+    let ms = ms_
+        .iter()
+        .zip(m_lens_.iter())
+        .map(|(&ptr, &m_len)| slice_from_c_bytes!(ptr, m_len));
+
+    let pks_lens_: &[size_t] = if len == 0 {
+        &[]
+    } else {
+        unsafe { slice::from_raw_parts(pks_len, len) }
+    };
+
+    let pks_ptrs: &[*const *mut PublicKey<Bls12>] = if len == 0 {
+        &[]
+    } else {
+        unsafe { slice::from_raw_parts(pks_ptr, len) }
+    };
+
+    let pks = unsafe {
+        pks_ptrs
+            .iter()
+            .zip(pks_lens_.iter())
+            .map(|(&ptr, &pk_len)| {
+                let pk_set: &[*mut PublicKey<Bls12>] = if pk_len == 0 {
+                    &[]
+                } else {
+                    slice::from_raw_parts(ptr, pk_len)
+                };
+                let res: Vec<PublicKey<Bls12>> = pk_set.iter().map(|pk| *from_ptr!(*pk)).collect();
```

Comment:
Ideally we would not be collecting things here, and the verify function would just take an iterator.

But because it is parallelized then this gets complex, and the types are complex, so it is not worth it.